### PR TITLE
Cleanup functions and fix None project

### PR DESCRIPTION
**Pull request by vmrob**

### CODE REVIEW

File: needy/library.py [hunk 1]
```diff
@@ -106,50 +106,58 @@ def build(self):
...
+                raise RuntimeError('unknown project type')
 
             unrecognized_configuration_keys = set(configuration.keys()) - project.configuration_keys() - self.additional_project_configuration_keys()
             if len(unrecognized_configuration_keys):
                 raise RuntimeError('unrecognized project configuration keys: {}'.format(', '.join(unrecognized_configuration_keys)))
 
             project.set_string_format_variables(**self.string_format_variables())
 
-            if not project:
-                raise RuntimeError('unknown project type')
```
Comment (vmrob) [inline needy/library.py]: This check appears to come a little bit too late--after `project.configuration_keys` and `project. set_string_format_variables` are called. Bumped it up a little.

File: needy/library.py [hunk 2]
```diff
@@ -106,50 +106,58 @@ def build(self):
...
 
-            if os.path.exists(build_directory):
+    def __clean(self):
+        configuration = self.project_configuration()
+        post_clean_commands = configuration['post-clean'] if 'post-clean' in configuration else []
+        with cd(self.project_root()):
+            for cmd in self.evaluate(post_clean_commands):
+                command(cmd)
+
+    def __actualize(self, project):
```
Comment (vmrob) [inline needy/library.py]: Very philosophical sounding, but I couldn't think of a better verb that covers setup, configure, and build.

File: needy/library.py [hunk 3]
```diff
@@ -106,50 +106,58 @@ def build(self):
 
         self.__log_env_overrides(env_overrides)
         with OverrideEnvironment(env_overrides):
-            post_clean_commands = configuration['post-clean'] if 'post-clean' in configuration else []
-            with cd(self.project_root()):
-                for cmd in self.evaluate(post_clean_commands):
-                    command(cmd)
+            self.__clean()
```
Comment (ccbrown) [inline needy/library.py]: I don't think this is the right name. The cleaning happens above. I'd suggest naming it `__post_clean`.

Comment (vmrob) [inline needy/library.py]: mmmm, yup. definitely true.

File: needy/library.py [hunk 4]
```diff
@@ -106,50 +106,58 @@ def build(self):
...
-            if os.path.exists(build_directory):
+    def __clean(self):
+        configuration = self.project_configuration()
+        post_clean_commands = configuration['post-clean'] if 'post-clean' in configuration else []
+        with cd(self.project_root()):
+            for cmd in self.evaluate(post_clean_commands):
+                command(cmd)
+
+    def __actualize(self, project):
+        build_directory = self.__prepare_build_directory()
```
Comment (ccbrown) [inline needy/library.py]: I'm not a fan of this convention... It's as if `__prepare_build_directory` is actually determining the build directory... but it doesn't. It just returns `self.build_directory()`. I'd prefer two separate invocations, maybe something like this:

```
build_directory=self.build_directory()
self.clean_directory(build_directory) # static and testable
```

File: needy/library.py (hunk 2 above)
Comment (ccbrown) [inline needy/library.py]: I think `__configure_and_build` would be good.

Comment (ccbrown) [inline needy/library.py]: Or maybe `__execute_build`. Or `__actualize`. Whatever's fine.

File: needy/library.py (hunk 4 above)
Comment (vmrob) [inline needy/library.py]: Agreed. Writing the tests now.

**Comment by ccbrown** [comment]
Only failure was a download timeout on a single job. Someone else's issue, but we should probably re-up that file if it's unreliable.

Lgtm.

**Pull request merged**